<!doctype html>
<html lang="en">

  <head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <link rel="stylesheet" href="https://unpkg.com/purecss@2.0.3/build/pure-min.css" integrity="sha384-cg6SkqEOCV1NbJoCu11+bm0NvBRc8IYLRGXkmNrqUBfTjmMYwNKPWBTIKyw9mHNJ" crossorigin="anonymous">
  <link rel="stylesheet" href="https://unpkg.com/purecss@2.0.3/build/grids-responsive-min.css">
  <link rel="stylesheet" href="/assets/styles.css"><!-- Begin Jekyll SEO tag v2.8.0 -->
<title>使用Matrix-eQTL完成cis-eQTL定位 | kkmonster</title>
<meta name="generator" content="Jekyll v3.10.0" />
<meta property="og:title" content="使用Matrix-eQTL完成cis-eQTL定位" />
<meta name="author" content="Feng kunjiang" />
<meta property="og:locale" content="en" />
<meta name="description" content="表达数量性状位点（eQTL）" />
<meta property="og:description" content="表达数量性状位点（eQTL）" />
<link rel="canonical" href="https://kunmonster.github.io//2024/10/25/%E4%BD%BF%E7%94%A8Matrix-eQTL%E5%AE%8C%E6%88%90cis-eQTL%E5%AE%9A%E4%BD%8D/" />
<meta property="og:url" content="https://kunmonster.github.io//2024/10/25/%E4%BD%BF%E7%94%A8Matrix-eQTL%E5%AE%8C%E6%88%90cis-eQTL%E5%AE%9A%E4%BD%8D/" />
<meta property="og:site_name" content="kkmonster" />
<meta property="og:type" content="article" />
<meta property="article:published_time" content="2024-10-25T21:05:00+08:00" />
<meta name="twitter:card" content="summary" />
<meta property="twitter:title" content="使用Matrix-eQTL完成cis-eQTL定位" />
<meta name="google-site-verification" content="xxxxx" />
<script type="application/ld+json">
{"@context":"https://schema.org","@type":"BlogPosting","author":{"@type":"Person","name":"Feng kunjiang","url":"https://github.com/kunmonster/"},"dateModified":"2024-10-25T21:05:00+08:00","datePublished":"2024-10-25T21:05:00+08:00","description":"表达数量性状位点（eQTL）","headline":"使用Matrix-eQTL完成cis-eQTL定位","mainEntityOfPage":{"@type":"WebPage","@id":"https://kunmonster.github.io//2024/10/25/%E4%BD%BF%E7%94%A8Matrix-eQTL%E5%AE%8C%E6%88%90cis-eQTL%E5%AE%9A%E4%BD%8D/"},"url":"https://kunmonster.github.io//2024/10/25/%E4%BD%BF%E7%94%A8Matrix-eQTL%E5%AE%8C%E6%88%90cis-eQTL%E5%AE%9A%E4%BD%8D/"}</script>
<!-- End Jekyll SEO tag -->
<link type="application/atom+xml" rel="alternate" href="https://kunmonster.github.io//feed.xml" title="kkmonster" />
</head>


  <body>

    <div class="container pure-g"><div class="sidebar-left pure-u-1 pure-u-md-1-4" style="background-color: rgb(40, 73, 77); color: rgb(255, 255, 255); background-image: url(/bg.jpg);"><header class="masthead">
  <div class="avatar"><img src="/avatar.jpg" class="avatar-image" alt="Feng kunjiang">
    </div>
  <div class="masthead-title">
    <a href="/" title="Home">kkmonster</a>
  </div>
  <div class="masthead-tagline">
    <small>生命中有万事的可能</small>
  </div><nav class="navigation">
    <ul class="navigation-list"><li class="navigation-item">
        <a onclick="sessionStorage.setItem('forceCheckScroll', 'true')" href="/">Blog</a>
      </li><li class="navigation-item">
        <a onclick="sessionStorage.setItem('forceCheckScroll', 'true')" href="/about/">About</a>
      </li></ul>
  </nav><div class="social pure-menu pure-menu-horizontal">
      <ul class="social-icons pure-menu-list">
      <li class="pure-menu-item">
          <a class="social-icon pure-menu-link" href="mailto://fengkunjiang@stu.ynu.edu.cn">
            <i class="fas fa-envelope" title="Email"></i>
          </a>
        </li><li class="pure-menu-item">
          <a class="social-icon pure-menu-link" href="https://github.com/kunmonster">
            <i class="fab fa-github" title="Github"></i>
          </a>
        </li>
      </ul>
    </div>
</header>
</div>

      <div class="content pure-u-1 pure-u-md-1-2"><main>
  <article class="post">
  <h1 class="post-title">使用Matrix-eQTL完成cis-eQTL定位</h1>
  <div class="post-meta"><time datetime="2024-10-25T21:05:00+08:00" itemprop="datePublished">25 Oct 2024</time></div>

  <h2 id="表达数量性状位点eqtl">表达数量性状位点（eQTL）</h2>
<hr />

<p>基因组上有些区域能导致某些基因的表达(mRNA)产生明显的变化，这些位置或者区域被称为表达数量性状位点(eQTL)，一般认为离目标基因1Mb位置以内的为顺式位点（local-eQTLs ， cis-eQTLs），与目标基因位置超过1Mb的成为反式位点（distant-eQTLs ， trans-eQTLs），主要通过线性回归对兴趣位点与目标基因表达量进行回归，并检验关系是否显著。</p>

<h2 id="使用matrix-eqtl进行大规模eqtl定位">使用Matrix eQTL进行大规模eQTL定位</h2>
<hr />

<p>本文只进行cis-eQTL和trans-eQTL定位，以下是本人真实研究数据在此进行举例并为自己提供笔记</p>

<h3 id="数据准备">数据准备</h3>
<hr />

<p>进行cis-eQTL和trans-eQTL定位需要准备以下数据</p>

<ul>
  <li>snp_genotype_matrix（基因分型矩阵，行为snp id，列为每个样本）(本例子中使用sv)</li>
</ul>

<blockquote>
  <ol>
    <li>snpid必须与下面的snp_loc中的id一样，并且sample_name的顺序必须和下面的expression_matrix中的sample_name顺序一致。</li>
    <li>并且matrix指的是R语言中的matrix，所以第一列就直接是值了，snpid应该设置为rowname。</li>
  </ol>
</blockquote>

<table>
  <thead>
    <tr>
      <th style="text-align: center">snpid</th>
      <th>sample_name-1</th>
      <th>sample_name-2</th>
      <th>sample_name-3</th>
      <th>……</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td style="text-align: center">1</td>
      <td>0（基因型）</td>
      <td>1（基因型）</td>
      <td>2（基因型）</td>
      <td>……</td>
    </tr>
  </tbody>
</table>

<ul>
  <li>expression_matrix（表达量矩阵，行为基因名或者基因ID,只要与下面的gene_pos中的一致即可，列为每个样本）</li>
</ul>

<blockquote>
  <ol>
    <li>注意表达量必须为数值类型的</li>
    <li>基因名字必须与下面的gene_pos中的基因名一致</li>
    <li>基因名字作为matrix的rowname</li>
  </ol>
</blockquote>

<table>
  <thead>
    <tr>
      <th style="text-align: center">gene_id</th>
      <th>sample_name-1</th>
      <th>sample_name-2</th>
      <th>sample_name-3</th>
      <th>……</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td style="text-align: center">gene_id</td>
      <td>该基因在该样本的表达量</td>
      <td>同左</td>
      <td>同左</td>
      <td>……</td>
    </tr>
  </tbody>
</table>

<ul>
  <li>gene_pos（基因相关信息）</li>
</ul>

<blockquote>
  <p>该文件应该以data.frame形式存入R对象中，并且列名应该严格按照下面的示例</p>
</blockquote>

<table>
  <thead>
    <tr>
      <th style="text-align: center">geneid</th>
      <th>chr</th>
      <th>left</th>
      <th>right</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td style="text-align: center">基因ID</td>
      <td>染色体号</td>
      <td>基因起点</td>
      <td>基因终点</td>
    </tr>
  </tbody>
</table>

<ul>
  <li>snp_loc （变异位点信息）</li>
</ul>

<blockquote>
  <p>同上，本内容也应在R中使用data.frame对象存放，列名如下</p>
</blockquote>

<table>
  <thead>
    <tr>
      <th style="text-align: center">snpid</th>
      <th>chr</th>
      <th>pos</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td style="text-align: center">变异ID</td>
      <td>染色体号</td>
      <td>变异位置</td>
    </tr>
  </tbody>
</table>

<h3 id="matrix-eqtl-api简介">Matrix eQTL api简介</h3>
<hr />
<p>matrix eQTL使用R编写，以快速进行大规模eQTL定位为人熟知，下面是本例要用的api以及主要参数解释</p>

<div class="language-R highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="w">
</span><span class="c1"># 本例子中主要使用Matrix_eQTL_main function</span><span class="w">
</span><span class="c1"># 其主要参数及解释如下</span><span class="w">

</span><span class="n">Matrix_eQTL_main</span><span class="p">(</span><span class="w">
        </span><span class="n">snps</span><span class="p">,</span><span class="w"> </span><span class="c1"># 基因型矩阵，需要SliceData对象,要求见数据准备部分,SliceData稍后介绍 </span><span class="w">
        </span><span class="n">gene</span><span class="p">,</span><span class="w"> </span><span class="c1"># 表达量矩阵，需要SliceData对象</span><span class="w">
        </span><span class="n">cvrt</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">SlicedData</span><span class="o">$</span><span class="n">new</span><span class="p">(),</span><span class="w"> </span><span class="c1"># 协变量，默认为空</span><span class="w">
        </span><span class="n">output_file_name</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">""</span><span class="p">,</span><span class="w"> </span><span class="c1"># 默认输出文件</span><span class="w">
        </span><span class="n">pvOutputThreshold</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1e-5</span><span class="p">,</span><span class="w"> </span><span class="c1"># 显著阈值,默认值10^(-5)</span><span class="w">
        </span><span class="n">useModel</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">modelLINEAR</span><span class="p">,</span><span class="w"> </span><span class="c1"># 模型选择,默认线性模型,还有两种其他模型,见Matrix eQTL文档</span><span class="w">
        </span><span class="n">errorCovariance</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">numeric</span><span class="p">(),</span><span class="w">  
        </span><span class="n">verbose</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">,</span><span class="w"> </span><span class="c1"># 过程日志输出</span><span class="w">
        </span><span class="n">output_file_name.cis</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">""</span><span class="p">,</span><span class="w"> </span><span class="c1"># cis-eQTL显著关联结果的输出文件 </span><span class="w">
        </span><span class="n">pvOutputThreshold.cis</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="c1"># cis-eQTL的显著值阈值</span><span class="w">
        </span><span class="n">snpspos</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">NULL</span><span class="p">,</span><span class="w"> </span><span class="c1"># data.frame,cis-eQTL和trans-eQTL需要,因为需要提供snp的位置</span><span class="w">
        </span><span class="n">genepos</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">NULL</span><span class="p">,</span><span class="w"> </span><span class="c1"># data.frame,同理在进行cis-eQTl和trans-eQTL时需要</span><span class="w">
        </span><span class="n">cisDist</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1e6</span><span class="p">,</span><span class="w"> </span><span class="c1"># cis-eQTL的距离，默认为1Mb范围内</span><span class="w">
        </span><span class="n">pvalue.hist</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">FALSE</span><span class="w">
        </span><span class="n">min.pv.by.genesnp</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">FALSE</span><span class="p">,</span><span class="w">
        </span><span class="n">noFDRsaveMemory</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">FALSE</span><span class="p">)</span><span class="w">

</span><span class="c1"># 注意如果不指定pvOutputThreshold.cis,snpspos,genepos将不会进行cis-eQTL和trans-eQTL分析,将会对所有候选位点和候选基因进行关联分析</span><span class="w">

</span><span class="c1"># 当 pvOutputThreshold=0, pvOutputThreshold.cis&gt;0时将只进行cis-eQTL分析</span><span class="w">
</span><span class="c1"># 当 pvOutputThreshold&gt;0,pvOutputThreshold.cis&gt;0时将进行cis-eQTL和trans-eQTl,其中trans-eQTL显著值使用pvOutputThreshold</span><span class="w">


</span><span class="c1"># 注意，前面在数据准备过程中,snp_genotype_matrix,expression_matrix被存成R矩阵,但是上面api要求输入SliceData对象</span><span class="w">
</span><span class="c1"># 所以需要转换成SliceData对象,SliceData类是Matrix eQTL中的一个用于矩阵的类,具体可以其帮助中查看</span><span class="w">

</span></code></pre></div></div>

<h3 id="具体实施过程">具体实施过程</h3>

<p>数据示例</p>

<p>gene_expression:读取进来的时候使用data.frame存储，后面要进行处理</p>

<p><img src="https://raw.githubusercontent.com/kunmonster/note_pic/main/note/202410240942849.png" alt="image.png" /></p>

<p>genotype:同上，以data.frame存储，后续要进行处理</p>

<p><img src="https://raw.githubusercontent.com/kunmonster/note_pic/main/note/202410240943149.png" alt="image.png" /></p>

<p>snp_pos:这个读取进来用data.frame存储即可，但是列名需要更换，第一列的ID应该与genotype中的ID一致
<img src="https://raw.githubusercontent.com/kunmonster/note_pic/main/note/202410240944479.png" alt="image.png" /></p>

<p>gene_pos:读取进来直接使用data.frame存储，gene_id列应与gene_exp中行名保持一致</p>

<p><img src="https://raw.githubusercontent.com/kunmonster/note_pic/main/note/202410240953497.png" alt="image.png" /></p>

<div class="language-R highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># 读取需要的文件</span><span class="w">
</span><span class="c1"># 基因位置</span><span class="w">
</span><span class="n">gene_pos</span><span class="o">&lt;-</span><span class="n">read.csv</span><span class="p">(</span><span class="s1">'path_to/gene_pos.csv'</span><span class="p">,</span><span class="n">header</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">)</span><span class="w">
</span><span class="c1"># 候选变异的位置信息</span><span class="w">
</span><span class="n">sv_loc</span><span class="o">&lt;-</span><span class="n">read.csv</span><span class="p">(</span><span class="s1">'path_to/sv_loc.csv'</span><span class="p">,</span><span class="n">header</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">)</span><span class="w">
</span><span class="n">genotype</span><span class="o">&lt;-</span><span class="n">read.csv</span><span class="p">(</span><span class="s1">'path_to/genotype.csv'</span><span class="p">,</span><span class="n">header</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">)</span><span class="w">
</span><span class="n">gene_expression</span><span class="o">&lt;-</span><span class="n">read.csv</span><span class="p">(</span><span class="s1">'path_to/expression_tpm.csv'</span><span class="p">,</span><span class="n">header</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">)</span><span class="w">


</span><span class="c1"># 过滤基因表达量低的基因,每个基因在所有的样本中有sample_percentage百分比的样本的表达量高于threshold才保留,并进行log转换</span><span class="w">
</span><span class="n">filter_low_expression_genes</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="k">function</span><span class="p">(</span><span class="n">expr_matrix</span><span class="p">,</span><span class="w"> </span><span class="n">threshold</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">sample_percentage</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.3</span><span class="p">)</span><span class="w"> </span><span class="p">{</span><span class="w">
  </span><span class="n">gene_names</span><span class="o">&lt;-</span><span class="n">expr_matrix</span><span class="p">[,</span><span class="m">1</span><span class="p">]</span><span class="w">
  </span><span class="n">expr_values</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">as.data.frame</span><span class="p">(</span><span class="n">lapply</span><span class="p">(</span><span class="n">expr_matrix</span><span class="p">[,</span><span class="w"> </span><span class="m">-1</span><span class="p">],</span><span class="w"> </span><span class="n">as.numeric</span><span class="p">))</span><span class="w">
  </span><span class="c1"># 计算超过阈值的样本数量阈值</span><span class="w">
  </span><span class="n">sample_threshold</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">ceiling</span><span class="p">(</span><span class="n">ncol</span><span class="p">(</span><span class="n">expr_values</span><span class="p">)</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="n">sample_percentage</span><span class="p">)</span><span class="w">
  </span><span class="c1"># 对每个基因，计算在样本中TPM&gt;=1的样本数</span><span class="w">
  </span><span class="n">filtered_matrix</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">expr_matrix</span><span class="p">[</span><span class="n">rowSums</span><span class="p">(</span><span class="n">expr_values</span><span class="w"> </span><span class="o">&gt;=</span><span class="w"> </span><span class="n">threshold</span><span class="p">)</span><span class="w"> </span><span class="o">&gt;=</span><span class="w"> </span><span class="n">sample_threshold</span><span class="p">,</span><span class="w"> </span><span class="p">]</span><span class="w">
  </span><span class="c1"># 对过滤后的表达矩阵进行 log2 转换 (log2(TPM + 1))  </span><span class="w">
  </span><span class="n">log_transformed_matrix</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">log2</span><span class="p">(</span><span class="n">filtered_matrix</span><span class="p">[,</span><span class="m">-1</span><span class="p">]</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="m">1</span><span class="p">)</span><span class="w">
  </span><span class="n">log_transformed_matrix</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">cbind</span><span class="p">(</span><span class="n">Gene</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">gene_names</span><span class="p">[</span><span class="n">rowSums</span><span class="p">(</span><span class="n">expr_values</span><span class="w"> </span><span class="o">&gt;=</span><span class="w"> </span><span class="n">threshold</span><span class="p">)</span><span class="w"> </span><span class="o">&gt;=</span><span class="w"> </span><span class="n">sample_threshold</span><span class="p">],</span><span class="w"> </span><span class="n">log_transformed_matrix</span><span class="p">)</span><span class="w">
  </span><span class="nf">return</span><span class="p">(</span><span class="n">log_transformed_matrix</span><span class="p">)</span><span class="w">
</span><span class="p">}</span><span class="w">


</span><span class="c1"># 检查基因位置和表达矩阵中基因的交集并返回</span><span class="w">
</span><span class="c1"># param: 表达矩阵,基因信息</span><span class="w">
</span><span class="n">check_the_cross</span><span class="o">&lt;-</span><span class="k">function</span><span class="p">(</span><span class="n">gene_exp</span><span class="p">,</span><span class="n">gene_pos</span><span class="p">){</span><span class="w">
  </span><span class="k">if</span><span class="p">(</span><span class="nf">is.null</span><span class="p">(</span><span class="n">gene_exp</span><span class="p">)){</span><span class="w">
    </span><span class="n">print</span><span class="p">(</span><span class="s2">"表达矩阵不存在"</span><span class="p">)</span><span class="w">
    </span><span class="nf">return</span><span class="p">(</span><span class="m">-1</span><span class="p">)</span><span class="w">
  </span><span class="p">}</span><span class="w">
  </span><span class="k">else</span><span class="w"> </span><span class="k">if</span><span class="p">(</span><span class="nf">is.null</span><span class="p">(</span><span class="n">gene_pos</span><span class="p">)){</span><span class="w">
    </span><span class="n">print</span><span class="p">(</span><span class="s2">"基因信息不存在"</span><span class="p">)</span><span class="w">
    </span><span class="nf">return</span><span class="p">(</span><span class="m">-1</span><span class="p">)</span><span class="w">
  </span><span class="p">}</span><span class="w">
  </span><span class="k">else</span><span class="p">{</span><span class="w">
    </span><span class="n">library</span><span class="p">(</span><span class="n">dplyr</span><span class="p">)</span><span class="w">
    </span><span class="n">result</span><span class="o">&lt;-</span><span class="n">filter_low_expression_genes</span><span class="p">(</span><span class="n">gene_exp</span><span class="p">)</span><span class="w">
    </span><span class="n">result</span><span class="o">&lt;-</span><span class="w"> </span><span class="n">result</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> 
      </span><span class="n">filter</span><span class="p">(</span><span class="n">result</span><span class="p">[[</span><span class="m">1</span><span class="p">]]</span><span class="w"> </span><span class="o">%in%</span><span class="w"> </span><span class="n">gene_pos</span><span class="p">[[</span><span class="m">1</span><span class="p">]])</span><span class="w">
    </span><span class="c1"># 处理表达量数据，去除表达量小的基因</span><span class="w">
    </span><span class="nf">return</span><span class="p">(</span><span class="n">result</span><span class="p">)</span><span class="w">
  </span><span class="p">}</span><span class="w">
</span><span class="p">}</span><span class="w">

</span><span class="c1"># 只保留兴趣基因的的表达量</span><span class="w">
</span><span class="n">gene_expression_cross</span><span class="o">&lt;-</span><span class="n">check_the_cross</span><span class="p">(</span><span class="n">gene_expression</span><span class="p">,</span><span class="n">gene_pos</span><span class="p">)</span><span class="w">
</span><span class="n">new_genotype</span><span class="o">&lt;-</span><span class="n">genotype</span><span class="w">
</span><span class="c1"># 将genotype中的缺失值替换成NA</span><span class="w">
</span><span class="n">new_genotype</span><span class="p">[</span><span class="n">new_genotype</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s2">"."</span><span class="p">]</span><span class="o">&lt;-</span><span class="kc">NA</span><span class="w">

</span><span class="c1"># 获取表达矩阵和基因型矩阵中样本的交集(有些样本没有表达数据，有些样本没有基因型数据)</span><span class="w">
</span><span class="n">common_names</span><span class="o">&lt;-</span><span class="n">intersect</span><span class="p">(</span><span class="nf">names</span><span class="p">(</span><span class="n">gene_expression_cross</span><span class="p">),</span><span class="nf">names</span><span class="p">(</span><span class="n">new_genotype</span><span class="p">))</span><span class="w">
</span><span class="n">gene_expression_cross</span><span class="o">&lt;-</span><span class="n">gene_expression_cross</span><span class="p">[,</span><span class="nf">c</span><span class="p">(</span><span class="s1">'Gene'</span><span class="p">,</span><span class="n">common_names</span><span class="p">)]</span><span class="w">
</span><span class="n">new_genotype</span><span class="o">&lt;-</span><span class="n">new_genotype</span><span class="p">[,</span><span class="nf">c</span><span class="p">(</span><span class="s1">'ID'</span><span class="p">,</span><span class="n">common_names</span><span class="p">)]</span><span class="w">

</span><span class="c1"># 按列名排序，确保两个data.frame列名序一样，这很重要</span><span class="w">
</span><span class="n">gene_expression_cross</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">gene_expression_cross</span><span class="p">[,</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s1">'Gene'</span><span class="p">,</span><span class="nf">names</span><span class="p">(</span><span class="n">gene_expression_cross</span><span class="p">)[</span><span class="m">-1</span><span class="p">][</span><span class="n">order</span><span class="p">(</span><span class="nf">names</span><span class="p">(</span><span class="n">gene_expression_cross</span><span class="p">)[</span><span class="m">-1</span><span class="p">])])]</span><span class="w">
</span><span class="n">new_genotype</span><span class="o">&lt;-</span><span class="n">new_genotype</span><span class="p">[,</span><span class="nf">c</span><span class="p">(</span><span class="s1">'ID'</span><span class="p">,</span><span class="nf">names</span><span class="p">(</span><span class="n">new_genotype</span><span class="p">)[</span><span class="m">-1</span><span class="p">][</span><span class="n">order</span><span class="p">(</span><span class="nf">names</span><span class="p">(</span><span class="n">new_genotype</span><span class="p">)[</span><span class="m">-1</span><span class="p">])])]</span><span class="w">


</span><span class="c1"># 由于matrix_eqtl需要特定类型的数据，所以将data.frame去掉列名，首列后，转换成matrix</span><span class="w">
</span><span class="c1"># 取出第一列作为后续矩阵的行名</span><span class="w">
</span><span class="n">row_name</span><span class="o">&lt;-</span><span class="n">gene_expression_cross</span><span class="o">$</span><span class="n">Gene</span><span class="w">
</span><span class="c1"># 去除第一列</span><span class="w">
</span><span class="n">gene_exp_matrix</span><span class="o">&lt;-</span><span class="n">gene_expression_cross</span><span class="p">[,</span><span class="m">-1</span><span class="p">]</span><span class="w">
</span><span class="c1"># 列名置空</span><span class="w">
</span><span class="n">colnames</span><span class="p">(</span><span class="n">gene_exp_matrix</span><span class="p">)</span><span class="o">&lt;-</span><span class="kc">NULL</span><span class="w">
</span><span class="c1"># 转换为矩阵</span><span class="w">
</span><span class="n">gene_exp_matrix</span><span class="o">&lt;-</span><span class="n">as.matrix</span><span class="p">(</span><span class="n">gene_exp_matrix</span><span class="p">)</span><span class="w">
</span><span class="c1"># 再给矩阵指定rowname</span><span class="w">
</span><span class="n">row.names</span><span class="p">(</span><span class="n">gene_exp_matrix</span><span class="p">)</span><span class="o">&lt;-</span><span class="n">row_name</span><span class="w">

</span><span class="c1"># 同样操作</span><span class="w">
</span><span class="n">row_name_genotype</span><span class="o">&lt;-</span><span class="n">new_genotype</span><span class="o">$</span><span class="n">ID</span><span class="w">
</span><span class="n">genotype_matrix</span><span class="o">&lt;-</span><span class="n">new_genotype</span><span class="p">[,</span><span class="m">-1</span><span class="p">]</span><span class="w">
</span><span class="n">colnames</span><span class="p">(</span><span class="n">genotype_matrix</span><span class="p">)</span><span class="o">&lt;-</span><span class="kc">NULL</span><span class="w">
</span><span class="n">genotype_matrix</span><span class="o">&lt;-</span><span class="n">as.matrix</span><span class="p">(</span><span class="n">genotype_matrix</span><span class="p">)</span><span class="w">

</span><span class="c1"># 将矩阵的值转换为数值</span><span class="w">
</span><span class="n">genotype_matrix</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">matrix</span><span class="p">(</span><span class="nf">as.numeric</span><span class="p">(</span><span class="n">genotype_matrix</span><span class="p">),</span><span class="w"> </span><span class="n">nrow</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">nrow</span><span class="p">(</span><span class="n">genotype_matrix</span><span class="p">),</span><span class="w"> </span><span class="n">ncol</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">ncol</span><span class="p">(</span><span class="n">genotype_matrix</span><span class="p">))</span><span class="w">
</span><span class="n">rownames</span><span class="p">(</span><span class="n">genotype_matrix</span><span class="p">)</span><span class="o">&lt;-</span><span class="n">row_name_genotype</span><span class="w">

</span><span class="c1"># 使得两data.frame列名满足输入要求</span><span class="w">
</span><span class="n">colnames</span><span class="p">(</span><span class="n">sv_loc</span><span class="p">)</span><span class="o">&lt;-</span><span class="nf">c</span><span class="p">(</span><span class="s1">'snpid'</span><span class="p">,</span><span class="s1">'chr'</span><span class="p">,</span><span class="s1">'pos'</span><span class="p">)</span><span class="w">
</span><span class="n">colnames</span><span class="p">(</span><span class="n">gene_pos</span><span class="p">)</span><span class="o">&lt;-</span><span class="nf">c</span><span class="p">(</span><span class="s1">'geneid'</span><span class="p">,</span><span class="s1">'chr'</span><span class="p">,</span><span class="s1">'left'</span><span class="p">,</span><span class="s1">'right'</span><span class="p">)</span><span class="w">

</span><span class="c1"># 确保gene_pos中的基因和表达矩阵中的基因一致（因为前面对表达矩阵进行了过滤，有些基因被滤除）</span><span class="w">
</span><span class="n">gene_pos</span><span class="o">&lt;-</span><span class="n">gene_pos</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> 
  </span><span class="n">filter</span><span class="p">(</span><span class="n">gene_pos</span><span class="o">$</span><span class="n">geneid</span><span class="w"> </span><span class="o">%in%</span><span class="w"> </span><span class="n">gene_expression_cross</span><span class="o">$</span><span class="n">Gene</span><span class="p">)</span><span class="w">
  
</span><span class="c1"># eqtl相关参数</span><span class="w">
</span><span class="n">cis_threshold</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="m">1e-5</span><span class="w">
</span><span class="n">trans_threshold</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="m">1e-5</span><span class="w">
</span><span class="n">cis_dist</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="m">1e6</span><span class="w">

</span><span class="n">pvOutputThreshold_cis</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">cis_threshold</span><span class="p">;</span><span class="w">
</span><span class="n">pvOutputThreshold_tra</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">trans_threshold</span><span class="p">;</span><span class="w">

</span><span class="n">cisDist</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">cis_dist</span><span class="w">

</span><span class="n">library</span><span class="p">(</span><span class="n">MatrixEQTL</span><span class="p">)</span><span class="w">

</span><span class="n">me</span><span class="o">&lt;-</span><span class="n">Matrix_eQTL_main</span><span class="p">(</span><span class="w">
  </span><span class="n">snps</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">SlicedData</span><span class="p">(</span><span class="n">genotype_matrix</span><span class="p">),</span><span class="w"> </span><span class="c1"># 因为SliceData可以使用矩阵直接构造对象，所以前面转换成矩阵后直接构造SliceData对象</span><span class="w">
  </span><span class="n">gene</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">SlicedData</span><span class="p">(</span><span class="n">gene_exp_matrix</span><span class="p">),</span><span class="w"> </span><span class="c1"># 同上</span><span class="w">
  </span><span class="n">pvOutputThreshold</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">pvOutputThreshold_tra</span><span class="p">,</span><span class="w">
  </span><span class="n">pvOutputThreshold.cis</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">pvOutputThreshold_cis</span><span class="p">,</span><span class="w">
  </span><span class="n">useModel</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">modelLINEAR</span><span class="p">,</span><span class="w">
  </span><span class="n">errorCovariance</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">numeric</span><span class="p">(),</span><span class="w">
  </span><span class="n">verbose</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">,</span><span class="w">
  </span><span class="n">snpspos</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">sv_loc</span><span class="p">,</span><span class="w">
  </span><span class="n">genepos</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">gene_pos</span><span class="p">,</span><span class="w">
  </span><span class="n">cisDist</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">cisDist</span><span class="p">,</span><span class="w">
  </span><span class="n">output_file_name</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"trans_eqtls.txt"</span><span class="w">
  </span><span class="n">output_file_name.cis</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"cis_eqtls.txt"</span><span class="p">,</span><span class="w">
  </span><span class="n">pvalue.hist</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"qqplot"</span><span class="p">,</span><span class="w">
  </span><span class="n">min.pv.by.genesnp</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">FALSE</span><span class="p">,</span><span class="w">
  </span><span class="n">noFDRsaveMemory</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">FALSE</span><span class="w">
</span><span class="p">)</span><span class="w">

</span></code></pre></div></div>

<h3 id="结果解读">结果解读</h3>
<hr />
<p>上述定位完成后结果将存放在me对象中，该对象结构如下</p>

<p><img src="https://raw.githubusercontent.com/kunmonster/note_pic/main/note/202410251307586.png" alt="image.png" /></p>

<p>其中trans与cis分别存放两种类型的eQTL位点,接下来我们挑选出FDR&lt;=0.05并且最显著的一个cis-eQTL
这里挑选出来，snp_id=68,gene_id=”LONRF3”的一组关联，我们将LONRF3的表达量列出，并且将其对应的snp genotype列出，对其做线性拟合，然后绘制箱线图展示关系</p>

<p><img src="https://raw.githubusercontent.com/kunmonster/note_pic/main/note/202410252058554.png" alt="image.png" /></p>

<p>从图中可以看出基因型的不同确实导致了表达量的变化，可能这个变异以某种机制作用于该基因的表达，具体还得后续分析。</p>

<h2 id="end">End</h2>
<hr />
<p>本文章仅记录自己使用Matrix-eQTL进行eQTL定位的过程，如有问题或者误导可联系我交流。</p>


  
    
      <div class="post-tags-section">
  <i class="post-tags-icon fas fa-tags"></i>
  <ul class="post-tags"><li>
        <a class="post-tag" href="/tags/#eqtl">eqtl</a></li></ul>
</div>

  

  
</article>


<aside class="related">
  <h2>Related posts</h2>
  <ul class="related-posts">
    
      <li>
        <a href="/2025/03/16/DeepsvFilter-%E5%9F%BA%E4%BA%8E%E6%B7%B1%E5%BA%A6%E5%AD%A6%E4%B9%A0%E5%AE%8C%E6%88%90%E7%BB%93%E6%9E%9C%E5%8F%98%E5%BC%82%E8%BF%87%E6%BB%A4/">
          DeepsvFilter:基于深度学习完成结果变异过滤
          <small><time datetime="2025-03-16T17:11:00+08:00">16 Mar 2025</time></small>
        </a>
      </li>
    
      <li>
        <a href="/2024/11/16/IGV%E5%8F%AF%E8%A7%86%E5%8C%96BAM%E6%96%87%E4%BB%B6%E5%B8%AE%E5%8A%A9%E7%90%86%E8%A7%A3%E7%BB%93%E6%9E%84%E5%8F%98%E5%BC%82/">
          IGV可视化BAM文件帮助理解结构变异
          <small><time datetime="2024-11-16T17:42:00+08:00">16 Nov 2024</time></small>
        </a>
      </li>
    
      <li>
        <a href="/2024/10/09/%E8%AE%B0%E5%BD%95windows%E7%89%88%E6%9C%ACIGV%E7%9A%84%E4%B8%80%E4%B8%AAbug%E7%9A%84%E4%BF%AE%E5%A4%8D%E8%BF%87%E7%A8%8B/">
          记录windows版本IGV的一个bug的修复过程
          <small><time datetime="2024-10-09T15:11:00+08:00">09 Oct 2024</time></small>
        </a>
      </li>
    
  </ul>
</aside>


</main>

<footer class="footer"><small>
    &copy; 2023&nbsp;-&nbsp;2025 <a href="https://github.com/kunmonster/">Feng kunjiang</a>. All rights reserved.
  </small>
</footer>
</div>
      <div class="sidebar-right pure-u-1 pure-u-md-1-4">
<div  class="toc-wrapper">
  <h2 class="toc-title">Table of Contents</h2>
    <nav class="toc-nav">
      <ul class="toc">
  <li><a href="#表达数量性状位点eqtl">表达数量性状位点（eQTL）</a></li>
  <li><a href="#使用matrix-eqtl进行大规模eqtl定位">使用Matrix eQTL进行大规模eQTL定位</a>
    <ul>
      <li><a href="#数据准备">数据准备</a></li>
      <li><a href="#matrix-eqtl-api简介">Matrix eQTL api简介</a></li>
      <li><a href="#具体实施过程">具体实施过程</a></li>
      <li><a href="#结果解读">结果解读</a></li>
    </ul>
  </li>
  <li><a href="#end">End</a></li>
</ul>

  </nav>
</div>

</div>
    </div>

    <script async src="https://use.fontawesome.com/releases/v5.0.12/js/all.js"></script><script>
  function strip(str, remove) {
    while (str.length > 0 && remove.indexOf(str.charAt(0)) != -1) {
      str = str.substr(1);
    }
    while (str.length > 0 && remove.indexOf(str.charAt(str.length - 1)) != -1) {
      str = str.substr(0, str.length - 1);
    }
    return str;
  }

  function scroll() {
    console.log('scroll');
    window.scrollTo({
      left: 0, 
      top: window.innerHeight,
      behavior: 'smooth'
    });
    sessionStorage.removeItem('forceCheckScroll');
  }

  const forceCheckScroll = sessionStorage.getItem('forceCheckScroll') === 'true';
  const checkScroll = strip(window.location.pathname, '/') !== strip('', '/');

  if (forceCheckScroll || checkScroll) {
    const maxWidth = "(max-width: 48rem)";
    const result = window.matchMedia(maxWidth);
    if (result.matches) {
      scroll();
    } else {
      result.addListener((match) => {
        if (match.media == maxWidth) {
          if (match.matches) {
            scroll();
          }
        }
      });
    }
  }
</script>
<script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
<script id="MathJax-script" async src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
</body>
</html>
