<!doctype html>
<html lang="en">

  <head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <link rel="stylesheet" href="https://unpkg.com/purecss@2.0.3/build/pure-min.css" integrity="sha384-cg6SkqEOCV1NbJoCu11+bm0NvBRc8IYLRGXkmNrqUBfTjmMYwNKPWBTIKyw9mHNJ" crossorigin="anonymous">
  <link rel="stylesheet" href="https://unpkg.com/purecss@2.0.3/build/grids-responsive-min.css">
  <link rel="stylesheet" href="/assets/styles.css">
  
  <script type="text/x-mathjax-config">
MathJax.Hub.Config({
    TeX: {
        equationNumbers: {
            autoNumber: "AMS"
        }
    },
    tex2jax: {
        inlineMath: [ ['$','$'] ],
        displayMath: [ ['$$','$$'] ],
        processEscapes: true,
    }
});
</script>
<script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML">
</script>

  <!-- Begin Jekyll SEO tag v2.8.0 -->
<title>Pangenome学习：A gentle introduction to pangenomics | kkmonster</title>
<meta name="generator" content="Jekyll v3.10.0" />
<meta property="og:title" content="Pangenome学习：A gentle introduction to pangenomics" />
<meta name="author" content="Feng kunjiang" />
<meta property="og:locale" content="en" />
<meta name="description" content="学习泛基因组的基本概念以及构建方法" />
<meta property="og:description" content="学习泛基因组的基本概念以及构建方法" />
<link rel="canonical" href="https://kunmonster.github.io//2025/11/05/Pangenome%E5%AD%A6%E4%B9%A0-A_gentle_introduction_to_pangenomics/" />
<meta property="og:url" content="https://kunmonster.github.io//2025/11/05/Pangenome%E5%AD%A6%E4%B9%A0-A_gentle_introduction_to_pangenomics/" />
<meta property="og:site_name" content="kkmonster" />
<meta property="og:type" content="article" />
<meta property="article:published_time" content="2025-11-05T16:00:00+08:00" />
<meta name="twitter:card" content="summary" />
<meta property="twitter:title" content="Pangenome学习：A gentle introduction to pangenomics" />
<meta name="google-site-verification" content="xxxxx" />
<script type="application/ld+json">
{"@context":"https://schema.org","@type":"BlogPosting","author":{"@type":"Person","name":"Feng kunjiang","url":"https://github.com/kunmonster/"},"dateModified":"2025-11-05T16:00:00+08:00","datePublished":"2025-11-05T16:00:00+08:00","description":"学习泛基因组的基本概念以及构建方法","headline":"Pangenome学习：A gentle introduction to pangenomics","mainEntityOfPage":{"@type":"WebPage","@id":"https://kunmonster.github.io//2025/11/05/Pangenome%E5%AD%A6%E4%B9%A0-A_gentle_introduction_to_pangenomics/"},"url":"https://kunmonster.github.io//2025/11/05/Pangenome%E5%AD%A6%E4%B9%A0-A_gentle_introduction_to_pangenomics/"}</script>
<!-- End Jekyll SEO tag -->
<link type="application/atom+xml" rel="alternate" href="https://kunmonster.github.io//feed.xml" title="kkmonster" />
</head>


  <body>

    <div class="container pure-g"><div class="sidebar-left pure-u-1 pure-u-md-1-4" style="background-color: rgb(40, 73, 77); color: rgb(255, 255, 255); background-image: url(/bg.jpg);"><header class="masthead">
  <div class="avatar"><img src="/avatar.jpg" class="avatar-image" alt="Feng kunjiang">
    </div>
  <div class="masthead-title">
    <a href="/" title="Home">kkmonster</a>
  </div>
  <div class="masthead-tagline">
    <small>生命中有万事的可能</small>
  </div><nav class="navigation">
    <ul class="navigation-list"><li class="navigation-item">
        <a onclick="sessionStorage.setItem('forceCheckScroll', 'true')" href="/">Blog</a>
      </li><li class="navigation-item">
        <a onclick="sessionStorage.setItem('forceCheckScroll', 'true')" href="/about/">About</a>
      </li></ul>
  </nav><div class="social pure-menu pure-menu-horizontal">
      <ul class="social-icons pure-menu-list">
      <li class="pure-menu-item">
          <a class="social-icon pure-menu-link" href="mailto://fengkunjiang@stu.ynu.edu.cn">
            <i class="fas fa-envelope" title="Email"></i>
          </a>
        </li><li class="pure-menu-item">
          <a class="social-icon pure-menu-link" href="https://github.com/kunmonster">
            <i class="fab fa-github" title="Github"></i>
          </a>
        </li>
      </ul>
    </div>
</header>
</div>

      <div class="content pure-u-1 pure-u-md-1-2"><main>
  <article class="post">
  <h1 class="post-title">Pangenome学习：A gentle introduction to pangenomics</h1>
  <div class="post-meta"><time datetime="2025-11-05T16:00:00+08:00" itemprop="datePublished">5 Nov 2025</time></div>

  <p>学习泛基因组的基本概念以及构建方法</p>

<h2 id="definition">Definition</h2>

<ul>
  <li>一个物种或者一个群体存在的自然的变异，一个群体或者物种所有的基因组信息</li>
  <li>使用生物信息学工具得到的泛基因组</li>
</ul>

<p>Core genome</p>

<p>sequence shared by all individuals</p>

<p>Accessory / Shell genome</p>

<p>Sequence present in some but not all individuals</p>

<p>Dispensable / Cloud genome</p>

<p>Sequence found in only one individual</p>

<h2 id="type-of-pangenomes">Type of pangenomes</h2>
<hr />

<h3 id="prensence-absence-variation-pangenome-pav">Prensence-absence variation pangenome (PAV)</h3>
<hr />

<p>关注基因的存在与否，不关注基因具体的位置，序列，以及多样性等</p>

<p>core genome ： 群体中每个个体都存在的基因，或者物种间每个物种都存在的基因</p>

<p>accessory genome： 只在群体中部分个体存在的基因或者物种间部分物种存在的基因</p>

<p><img src="https://raw.githubusercontent.com/kunmonster/note_pic/main/note/202511051557045.png" alt="image.png" /></p>

<h4 id="construction">Construction</h4>

<p>homolog-based strategy</p>

<p>对每个样本进行de novo组装，组装后提取基因组中注释到的编码基因，并且在不同样本间根据基因序列相似性聚类，某个类别中如果包含所有的样本那么就认定为Core gene，否则认定为Accessory genes，通常用于原核生物，原核生物基因无内含子，比对，聚类成本较小。</p>

<p>map-to-pan strategy</p>

<p>将whole-genome sequencing reads比对到经过注释的具有代表性的泛基因组</p>
<ul>
  <li>测序深度最低10X</li>
  <li>基因外显子至少5%的区域被至少1个reads覆盖，被认为基因存在的标志</li>
</ul>

<p>softcore
当有一个样本中的Core gene没能被正确识别到的时候，那么整个Core gene合集可能就没有这个基因，为了防止这种情况，有些研究提出这个概念，允许在小部分样本中没识别到的基因为softcore</p>

<h3 id="representative-sequence-pangenome">Representative sequence pangenome</h3>
<hr />

<p>使用尽可能少同源序列，并尽可能多的表示基因组多样性，通常由一个参考基因组以及其他非冗余参考序列组成(NRR, nonredundant reference), 其中NRR包含只在某一个成员中出现，但是不在reference中出现的</p>

<p><img src="https://raw.githubusercontent.com/kunmonster/note_pic/main/note/202511051557915.png" alt="image.png" /></p>
<h4 id="construction-1">Construction</h4>

<p>将在参考基因组中不存在的基因组区域作为额外contig来构建</p>

<ul>
  <li>metagenome-like assembly of unaligned reads （低于10X的大样本测序数据首选）</li>
</ul>

<p><img src="https://raw.githubusercontent.com/kunmonster/note_pic/main/note/202511051502140.png" alt="image.png" /></p>

<p>先将所有样本reads比对到参考基因组，没比对上的所有reads一起进行de novo assembly ，形成contig</p>

<ul>
  <li>independent assembly of unaligned reads</li>
</ul>

<p><img src="https://raw.githubusercontent.com/kunmonster/note_pic/main/note/202511051502072.png" alt="image.png" /></p>

<p>先将每个样本reads比对到参考基因组，没比对上的按样本进行de novo组装，将组装好的contig根据序列相似度进行聚类，从每个聚类中选出一个作为参考基因组的补充（需要至少10x的read coverage）</p>

<ul>
  <li>iterative assembly of unaligned reads</li>
</ul>

<p><img src="https://raw.githubusercontent.com/kunmonster/note_pic/main/note/202511051505870.png" alt="image.png" /></p>

<p>迭代式创建，单个样本先进行比对，然后将未比对的reads进行de novo assembly，用于扩充参考基因组，然后更新的参考基因组再用于其他样本。后续样本就先比对到更新的参考基因组，然后扩充参考基因组，一直迭代。</p>

<ul>
  <li>independent whole-genome assembly</li>
</ul>

<p><img src="https://raw.githubusercontent.com/kunmonster/note_pic/main/note/202511051506422.png" alt="image.png" /></p>

<p>先将每个样本的reads进行de novo assembly至contig水平，然后将contig比对到参考基因组，所有样本未能比对至参考基因组的contigs进行聚类，从聚类结果中挑选每个类别中最长的序列用于扩充参考基因组</p>

<h3 id="pangenome-graph-graphic-pangenome">pangenome graph (graphic pangenome)</h3>
<hr />
<p>可以是面向序列的和面向基因的，展示群体中变异和基因的位置关系。</p>

<p><img src="https://raw.githubusercontent.com/kunmonster/note_pic/main/note/202511051526235.png" alt="image.png" /></p>

<h4 id="construction-2">Construction</h4>

<ul>
  <li>Predetermined variants</li>
</ul>

<p><img src="https://raw.githubusercontent.com/kunmonster/note_pic/main/note/202511051536370.png" alt="image.png" /></p>

<p>需要一个参考基因组，以及包含已有变异的VCF文件，后来新添加的变异作为分支节点添加到图上，形成一个有向无环图</p>

<ul>
  <li>Multiple sequence alignment</li>
</ul>

<p><img src="https://raw.githubusercontent.com/kunmonster/note_pic/main/note/202511051536320.png" alt="image.png" /></p>

<p>多序列比对，直接将序列两两比对进行图形基因组构建，这种方法需要大量的计算资源，但是对于定相的基因组十分有用(Phased genome)</p>

<ul>
  <li>De Bruijn graphs</li>
</ul>

<p><img src="https://raw.githubusercontent.com/kunmonster/note_pic/main/note/202511051536087.png" alt="image.png" /></p>

<p>将reads分成k-mers，然后根据k-1(k为kmer的长度)的重叠来进行图构建，kmer为图节点</p>

<h4 id="applications">Applications</h4>

<ul>
  <li>变异检测与分型（variants calling and genotyping）</li>
  <li>单倍型推断 (Haplotype inference; phasing of assemblies)</li>
</ul>

<h2 id="files-and-formats">Files and formats</h2>

<table>
  <thead>
    <tr>
      <th style="text-align: center"> </th>
      <th>Linear</th>
      <th>Pangenomic</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td style="text-align: center">Reference genome</td>
      <td>.fa</td>
      <td>.gfa</td>
    </tr>
    <tr>
      <td style="text-align: center">Indexing</td>
      <td>.bwt</td>
      <td>.gbwt</td>
    </tr>
    <tr>
      <td style="text-align: center">Alignment</td>
      <td>.bam / .paf</td>
      <td>.gam  / .gaf</td>
    </tr>
  </tbody>
</table>

<h2 id="reference">Reference</h2>
<hr />
<p>Chelsea A Matthews, Nathan S Watson-Haigh, Rachel A Burton, Anna E Sheppard, A gentle introduction to pangenomics, <em>Briefings in Bioinformatics</em>, Volume 25, Issue 6, November 2024, bbae588, <a href="https://doi.org/10.1093/bib/bbae588">https://doi.org/10.1093/bib/bbae588</a></p>


  
    
      <div class="post-tags-section">
  <i class="post-tags-icon fas fa-tags"></i>
  <ul class="post-tags"><li>
        <a class="post-tag" href="/tags/#paper">paper</a></li><li>
        <a class="post-tag" href="/tags/#pangenome">pangenome</a></li></ul>
</div>

  

  
</article>


<aside class="related">
  <h2>Related posts</h2>
  <ul class="related-posts">
    
      <li>
        <a href="/2025/09/14/CSV-Filter%E6%95%B0%E6%8D%AE%E5%87%86%E5%A4%87%E8%A7%A3%E8%AF%BB/">
          CSV-Filter数据准备解读
          <small><time datetime="2025-09-14T21:20:00+08:00">14 Sep 2025</time></small>
        </a>
      </li>
    
      <li>
        <a href="/2025/04/14/Circos_draw_histogram/">
          使用circos画变异分布直方图
          <small><time datetime="2025-04-14T22:00:00+08:00">14 Apr 2025</time></small>
        </a>
      </li>
    
      <li>
        <a href="/2025/04/11/%E4%BD%BF%E7%94%A8SV%E5%AE%8C%E6%88%90%E4%B8%BB%E6%88%90%E5%88%86%E5%88%86%E6%9E%90/">
          使用SV完成主成分分析
          <small><time datetime="2025-04-11T20:30:00+08:00">11 Apr 2025</time></small>
        </a>
      </li>
    
  </ul>
</aside>


</main>

<footer class="footer"><small>
    &copy; 2023&nbsp;-&nbsp;2025 <a href="https://github.com/kunmonster/">Feng kunjiang</a>. All rights reserved.
  </small>
</footer>
</div>
      <div class="sidebar-right pure-u-1 pure-u-md-1-4">
<div  class="toc-wrapper">
  <h2 class="toc-title">Table of Contents</h2>
    <nav class="toc-nav">
      <ul class="toc">
  <li><a href="#definition">Definition</a></li>
  <li><a href="#type-of-pangenomes">Type of pangenomes</a>
    <ul>
      <li><a href="#prensence-absence-variation-pangenome-pav">Prensence-absence variation pangenome (PAV)</a></li>
      <li><a href="#representative-sequence-pangenome">Representative sequence pangenome</a></li>
      <li><a href="#pangenome-graph-graphic-pangenome">pangenome graph (graphic pangenome)</a></li>
    </ul>
  </li>
  <li><a href="#files-and-formats">Files and formats</a></li>
  <li><a href="#reference">Reference</a></li>
</ul>

  </nav>
</div>

</div>
    </div>

    <script async src="https://use.fontawesome.com/releases/v5.0.12/js/all.js"></script><script>
  function strip(str, remove) {
    while (str.length > 0 && remove.indexOf(str.charAt(0)) != -1) {
      str = str.substr(1);
    }
    while (str.length > 0 && remove.indexOf(str.charAt(str.length - 1)) != -1) {
      str = str.substr(0, str.length - 1);
    }
    return str;
  }

  function scroll() {
    console.log('scroll');
    window.scrollTo({
      left: 0, 
      top: window.innerHeight,
      behavior: 'smooth'
    });
    sessionStorage.removeItem('forceCheckScroll');
  }

  const forceCheckScroll = sessionStorage.getItem('forceCheckScroll') === 'true';
  const checkScroll = strip(window.location.pathname, '/') !== strip('', '/');

  if (forceCheckScroll || checkScroll) {
    const maxWidth = "(max-width: 48rem)";
    const result = window.matchMedia(maxWidth);
    if (result.matches) {
      scroll();
    } else {
      result.addListener((match) => {
        if (match.media == maxWidth) {
          if (match.matches) {
            scroll();
          }
        }
      });
    }
  }
</script>
</body>
</html>
