<!doctype html>
<html lang="en">

  <head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <link rel="stylesheet" href="https://unpkg.com/purecss@2.0.3/build/pure-min.css" integrity="sha384-cg6SkqEOCV1NbJoCu11+bm0NvBRc8IYLRGXkmNrqUBfTjmMYwNKPWBTIKyw9mHNJ" crossorigin="anonymous">
  <link rel="stylesheet" href="https://unpkg.com/purecss@2.0.3/build/grids-responsive-min.css">
  <link rel="stylesheet" href="/assets/styles.css">
  
  <script type="text/x-mathjax-config">
MathJax.Hub.Config({
    TeX: {
        equationNumbers: {
            autoNumber: "AMS"
        }
    },
    tex2jax: {
        inlineMath: [ ['$','$'] ],
        displayMath: [ ['$$','$$'] ],
        processEscapes: true,
    }
});
</script>
<script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML">
</script>

  <!-- Begin Jekyll SEO tag v2.8.0 -->
<title>CSV-Filter数据准备解读 | kkmonster</title>
<meta name="generator" content="Jekyll v3.10.0" />
<meta property="og:title" content="CSV-Filter数据准备解读" />
<meta name="author" content="Feng kunjiang" />
<meta property="og:locale" content="en" />
<meta name="description" content="该研究基于深度学习开发了结构变异过滤工具，针对short-read 和 long-read方法检测的结构变异位点，本博客主要关注方法和材料部分。" />
<meta property="og:description" content="该研究基于深度学习开发了结构变异过滤工具，针对short-read 和 long-read方法检测的结构变异位点，本博客主要关注方法和材料部分。" />
<link rel="canonical" href="https://kunmonster.github.io//2025/09/14/CSV-Filter%E6%95%B0%E6%8D%AE%E5%87%86%E5%A4%87%E8%A7%A3%E8%AF%BB/" />
<meta property="og:url" content="https://kunmonster.github.io//2025/09/14/CSV-Filter%E6%95%B0%E6%8D%AE%E5%87%86%E5%A4%87%E8%A7%A3%E8%AF%BB/" />
<meta property="og:site_name" content="kkmonster" />
<meta property="og:type" content="article" />
<meta property="article:published_time" content="2025-09-14T21:20:00+08:00" />
<meta name="twitter:card" content="summary" />
<meta property="twitter:title" content="CSV-Filter数据准备解读" />
<meta name="google-site-verification" content="xxxxx" />
<script type="application/ld+json">
{"@context":"https://schema.org","@type":"BlogPosting","author":{"@type":"Person","name":"Feng kunjiang","url":"https://github.com/kunmonster/"},"dateModified":"2025-09-14T21:20:00+08:00","datePublished":"2025-09-14T21:20:00+08:00","description":"该研究基于深度学习开发了结构变异过滤工具，针对short-read 和 long-read方法检测的结构变异位点，本博客主要关注方法和材料部分。","headline":"CSV-Filter数据准备解读","mainEntityOfPage":{"@type":"WebPage","@id":"https://kunmonster.github.io//2025/09/14/CSV-Filter%E6%95%B0%E6%8D%AE%E5%87%86%E5%A4%87%E8%A7%A3%E8%AF%BB/"},"url":"https://kunmonster.github.io//2025/09/14/CSV-Filter%E6%95%B0%E6%8D%AE%E5%87%86%E5%A4%87%E8%A7%A3%E8%AF%BB/"}</script>
<!-- End Jekyll SEO tag -->
<link type="application/atom+xml" rel="alternate" href="https://kunmonster.github.io//feed.xml" title="kkmonster" />
</head>


  <body>

    <div class="container pure-g"><div class="sidebar-left pure-u-1 pure-u-md-1-4" style="background-color: rgb(40, 73, 77); color: rgb(255, 255, 255); background-image: url(/bg.jpg);"><header class="masthead">
  <div class="avatar"><img src="/avatar.jpg" class="avatar-image" alt="Feng kunjiang">
    </div>
  <div class="masthead-title">
    <a href="/" title="Home">kkmonster</a>
  </div>
  <div class="masthead-tagline">
    <small>生命中有万事的可能</small>
  </div><nav class="navigation">
    <ul class="navigation-list"><li class="navigation-item">
        <a onclick="sessionStorage.setItem('forceCheckScroll', 'true')" href="/">Blog</a>
      </li><li class="navigation-item">
        <a onclick="sessionStorage.setItem('forceCheckScroll', 'true')" href="/about/">About</a>
      </li></ul>
  </nav><div class="social pure-menu pure-menu-horizontal">
      <ul class="social-icons pure-menu-list">
      <li class="pure-menu-item">
          <a class="social-icon pure-menu-link" href="mailto://fengkunjiang@stu.ynu.edu.cn">
            <i class="fas fa-envelope" title="Email"></i>
          </a>
        </li><li class="pure-menu-item">
          <a class="social-icon pure-menu-link" href="https://github.com/kunmonster">
            <i class="fab fa-github" title="Github"></i>
          </a>
        </li>
      </ul>
    </div>
</header>
</div>

      <div class="content pure-u-1 pure-u-md-1-2"><main>
  <article class="post">
  <h1 class="post-title">CSV-Filter数据准备解读</h1>
  <div class="post-meta"><time datetime="2025-09-14T21:20:00+08:00" itemprop="datePublished">14 Sep 2025</time></div>

  <p>该研究基于深度学习开发了结构变异过滤工具，针对short-read 和 long-read方法检测的结构变异位点，本博客主要关注方法和材料部分。</p>

<p>该研究也将结构变异过滤问题转化为图像二分类问题。那么首要问题还是如何将结构变异转化为图像，以及negative training sample，与其他类似研究一样，该研究同样基于CIGAR比对信息来编码图像，并将图像编码拆解成：</p>
<ol>
  <li>Structural variant sites selecting;</li>
  <li>Reads selecting;</li>
  <li>Image coding。</li>
</ol>

<p>下文将解析作者的处理。</p>

<h3 id="negative-training-samples-preparing">Negative training samples preparing</h3>
<hr />
<p><img src="https://raw.githubusercontent.com/kunmonster/note_pic/main/note/202509141707109.png" alt="image.png" /></p>

<p>作者通过分析结构变的分布，发现变异的长度符合泊松分布。作者计算变异长度的平均值，方差并计算这两者的调和平均数。将调和平均数作为密度函数的参数，遵循该分布，生成negative samples。</p>

<p><img src="https://raw.githubusercontent.com/kunmonster/note_pic/main/note/202509141956013.png" alt="image.png" /></p>

<p>随后迭代生成负样本，如果生成的负样本和相邻的样本有超过一半的重合，那么应舍弃并重新生成。在完成迭代后，作者将产生的变异标准化。</p>

<p>图中，$\lambda_{sv}$ 应该truth set的sv长度的均值和方差的调和平均数</p>

<h3 id="sites-locating">Sites locating</h3>
<hr />

<h3 id="reads-selecting">Reads selecting</h3>
<hr />
<p><img src="https://raw.githubusercontent.com/kunmonster/note_pic/main/note/202509141638127.png" alt="image.png" /></p>

<p>如图所示，在确定变异位点后，围绕变异位点进行reads选取。</p>

<p>从图a可以看出，作者在断点周围向上游和下游都扩展了一段extension，所有覆盖整个范围的reads将会被选择。</p>

<h3 id="image-coding">Image coding</h3>
<hr />

<p><img src="https://raw.githubusercontent.com/kunmonster/note_pic/main/note/202509141704588.png" alt="CSV-Filter-img-coding.png" /></p>

<p>在选取reads后，维护两个列表用于存储所有reads的起点和终点，确认所有reads起点中最小的，命名为：$b_{min}$ ，  迭代所有reads，计算当前reads起点与$b_{min}$的偏移，确定当前reads编码的范围 ($b_{i} - b_{min},e_{i}$)。</p>

<p>在填充像素时，针对CIGAR中的”N”,”P”,”H”,”=”,”X” 以及偏移部分，填充0，上述reads编码范围部分使用kernal_cigar函数填充。</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="k">def</span> <span class="nf">kernel_cigar</span><span class="p">(</span><span class="n">read</span><span class="p">,</span> <span class="n">ref_min</span><span class="p">,</span> <span class="n">ref_max</span><span class="p">,</span> <span class="n">cigar_resize</span><span class="p">,</span> <span class="n">zoom</span><span class="p">):</span>
	<span class="n">cigars_img</span> <span class="o">=</span> <span class="n">torch</span><span class="p">.</span><span class="n">zeros</span><span class="p">([</span><span class="mi">1</span><span class="p">,</span> <span class="nb">int</span><span class="p">((</span><span class="n">ref_max</span> <span class="o">-</span> <span class="n">ref_min</span><span class="p">)</span> <span class="o">/</span> <span class="n">zoom</span><span class="p">)])</span>
	<span class="n">max_terminal</span> <span class="o">=</span> <span class="n">read</span><span class="p">.</span><span class="n">reference_start</span> <span class="o">-</span> <span class="n">ref_min</span>
</code></pre></div></div>

<h3 id="dataset">Dataset</h3>
<hr />

<p>该研究主要使用两个样本(HG002, NA12878)的相关文件用于训练</p>

<h4 id="hg002">HG002</h4>

<ul>
  <li>Third-generation data</li>
</ul>

<p>三代测序数据使用minimap2, pbmm2, NGMLR比对到参考基因组 <a href="https://ftp.ensembl.org/pub/release-75/fasta/homo_sapiens/dna/Homo_sapiens.GRCh37.75.dna.primary_assembly.fa.gz">GRCh37</a></p>

<p>测序数据包括</p>

<p><a href="https://ftp-trace.ncbi.nlm.nih.gov/ReferenceSamples/giab/data/AshkenazimTrio/HG002_NA24385_son/PacBio_MtSinai_NIST/">PacBio 70x(CLR)</a></p>

<p><a href="https://ftp-trace.ncbi.nlm.nih.gov/ReferenceSamples/giab/data/AshkenazimTrio/HG002_NA24385_son/PacBio_CCS_15kb_20kb_chemistry2/reads/">PacBio CCs 15kb_20kb chemistry2(HIFI)</a></p>

<p><a href="ftp://ftp-trace.ncbi.nlm.nih.gov/ReferenceSamples/giab/data/AshkenazimTrio/HG002_NA24385_son/Ultralong_OxfordNanopore/guppy-V3.2.4_2020-01-22/HG002_ONT-UL_GIAB_20200122.fastq.gz">Oxford Nanopore ultralong(guppy-V3.2.4_2020-01-22)</a></p>

<ul>
  <li>SV</li>
</ul>

<p>该研究选用了GIAB(Genome in a Bottle)的benchmark数据集-<a href="https://ftp-trace.ncbi.nlm.nih.gov/ReferenceSamples/giab/data/AshkenazimTrio/analysis/NIST_SVs_Integration_v0.6/">HG002_SVs_Tier1_v0.6</a>，该数据集包括4,199个deletions与5,442个insertions。</p>

<ul>
  <li>bam</li>
</ul>

<p>二代数据该研究使用来自HG002的illumina 300X，参考基因组为<a href="https://ftp-trace.ncbi.nih.gov/1000genomes/ftp/technical/reference/phase2_reference_assembly_sequence/hs37d5.fa.gz">hs37d5</a>。<a href="https://ftp-trace.ncbi.nlm.nih.gov/giab/ftp/data/AshkenazimTrio/HG002_NA24385_son/NIST_HiSeq_HG002_Homogeneity-10953946/NHGRI_Illumina300X_AJtrio_novoalign_bams/HG002.hs37d5.60x.1.bam">HG002.hs37d5.60x.1.bam</a></p>

<h4 id="na12878">NA12878</h4>

<ul>
  <li>SV</li>
</ul>

<p>有研究对NA12878的三代测序数据使用三种方法检测了结构变异，包括INS,DEL. <a href="https://ftp.ncbi.nlm.nih.gov/giab/ftp/data/NA12878/NA12878_PacBio_MtSinai/NA12878.sorted.vcf.gz">vcf-file</a></p>

<p>同时该样本还有基于二代测序数据的结构变异数据集. <a href="ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/phase3/integrated_sv_map/supporting/GRCh38_positions/ALL.wgs.mergedSV.v8.20130502.svs.genotypes.GRCh38.vcf.gz.">vcf-file</a></p>

<ul>
  <li>bam</li>
</ul>

<p>PacBio 测序数据比对到hg19的bam文件. <a href="https://ftp.ncbi.nlm.nih.gov/giab/ftp/data/NA12878/NA12878_PacBio_MtSinai/sorted_final_merged.bam">bam_file</a>
illumina测序数据比对到GRCh38的bam文件. <a href="ftp://ftp.sra.ebi.ac.uk/vol1/run/ERR194/ERR194147/NA12878_S1.bam">bam_file</a></p>

<ul>
  <li>reference</li>
</ul>

<p><a href="http://hgdownload.cse.ucsc.edu/goldenPath/hg19/bigZips/chromFa.tar.gz">hg19</a></p>

<p><a href="https://ftp-trace.ncbi.nih.gov/1000genomes/ftp/technical/reference/GRCh38_reference_genome/GRCh38_full_analysis_set_plus_decoy_hla.fa">GRCh38</a></p>

<h4 id="t2t-chm13">T2T-CHM13</h4>

<p>该数据集用于验证，该数据集没有对应的gold standard dataset, 作者使用Dipcall(v0.3)对CHM13的端到端版本进行变异检测，将大于50bp且位于高可信区域的变异视为 ground truth</p>

<p><a href="https://github.com/marbl/CHM13">T2T-CHM13</a></p>

<h4 id="ashkenazim-trio">Ashkenazim Trio</h4>

<p>作者还使用HG002, HG003, HG004统计CIGAR字符串中各个操作的比例. <a href="https://ftp.ncbi.nih.gov/giab/ftp/data/AshkenazimTrio/">Archive data for the Ashkenazim Trio</a></p>

<h3 id="reference">Reference</h3>
<hr />
<p>[1] Xia Z, Xiang W, Wang Q, Li X, Li Y, Gao J, et al. CSV-Filter: a deep learning-based comprehensive structural variant filtering method for both short and long reads. Bioinformatics. 2024;40(9).</p>

<p>[2] Xiang W, Cui Y, Yang Y, Zhang A, Ji B, Peng S, editors. MSVF: Multi-task Structure Variation Filter with Transfer Learning in High-throughput Sequencing. 2022 IEEE International Conference on Bioinformatics and Biomedicine (BIBM); 2022 6-8 Dec. 2022.</p>

<p>[3] Zook JM, Hansen NF, Olson ND, Chapman L, Mullikin JC, Xiao C, et al. A robust benchmark for detection of germline large deletions and insertions. Nat Biotechnol. 2020;38(11):1347-55.</p>

<p>……</p>

<p>本博客仅作为论文阅读笔记，学习过程的记录！</p>


  
    
      <div class="post-tags-section">
  <i class="post-tags-icon fas fa-tags"></i>
  <ul class="post-tags"><li>
        <a class="post-tag" href="/tags/#sv">sv</a></li><li>
        <a class="post-tag" href="/tags/#deeplearning">deeplearning</a></li></ul>
</div>

  

  
</article>


<aside class="related">
  <h2>Related posts</h2>
  <ul class="related-posts">
    
      <li>
        <a href="/2025/11/05/Pangenome%E5%AD%A6%E4%B9%A0-A_gentle_introduction_to_pangenomics/">
          Pangenome学习：A gentle introduction to pangenomics
          <small><time datetime="2025-11-05T16:00:00+08:00">05 Nov 2025</time></small>
        </a>
      </li>
    
      <li>
        <a href="/2025/04/14/Circos_draw_histogram/">
          使用circos画变异分布直方图
          <small><time datetime="2025-04-14T22:00:00+08:00">14 Apr 2025</time></small>
        </a>
      </li>
    
      <li>
        <a href="/2025/04/11/%E4%BD%BF%E7%94%A8SV%E5%AE%8C%E6%88%90%E4%B8%BB%E6%88%90%E5%88%86%E5%88%86%E6%9E%90/">
          使用SV完成主成分分析
          <small><time datetime="2025-04-11T20:30:00+08:00">11 Apr 2025</time></small>
        </a>
      </li>
    
  </ul>
</aside>


</main>

<footer class="footer"><small>
    &copy; 2023&nbsp;-&nbsp;2025 <a href="https://github.com/kunmonster/">Feng kunjiang</a>. All rights reserved.
  </small>
</footer>
</div>
      <div class="sidebar-right pure-u-1 pure-u-md-1-4">
<div  class="toc-wrapper">
  <h2 class="toc-title">Table of Contents</h2>
    <nav class="toc-nav">
      <ul class="toc">
  <li><a href="#negative-training-samples-preparing">Negative training samples preparing</a></li>
  <li><a href="#sites-locating">Sites locating</a></li>
  <li><a href="#reads-selecting">Reads selecting</a></li>
  <li><a href="#image-coding">Image coding</a></li>
  <li><a href="#dataset">Dataset</a></li>
  <li><a href="#reference">Reference</a></li>
</ul>

  </nav>
</div>

</div>
    </div>

    <script async src="https://use.fontawesome.com/releases/v5.0.12/js/all.js"></script><script>
  function strip(str, remove) {
    while (str.length > 0 && remove.indexOf(str.charAt(0)) != -1) {
      str = str.substr(1);
    }
    while (str.length > 0 && remove.indexOf(str.charAt(str.length - 1)) != -1) {
      str = str.substr(0, str.length - 1);
    }
    return str;
  }

  function scroll() {
    console.log('scroll');
    window.scrollTo({
      left: 0, 
      top: window.innerHeight,
      behavior: 'smooth'
    });
    sessionStorage.removeItem('forceCheckScroll');
  }

  const forceCheckScroll = sessionStorage.getItem('forceCheckScroll') === 'true';
  const checkScroll = strip(window.location.pathname, '/') !== strip('', '/');

  if (forceCheckScroll || checkScroll) {
    const maxWidth = "(max-width: 48rem)";
    const result = window.matchMedia(maxWidth);
    if (result.matches) {
      scroll();
    } else {
      result.addListener((match) => {
        if (match.media == maxWidth) {
          if (match.matches) {
            scroll();
          }
        }
      });
    }
  }
</script>
</body>
</html>
