<!doctype html>
<html lang="en">

  <head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <link rel="stylesheet" href="https://unpkg.com/purecss@2.0.3/build/pure-min.css" integrity="sha384-cg6SkqEOCV1NbJoCu11+bm0NvBRc8IYLRGXkmNrqUBfTjmMYwNKPWBTIKyw9mHNJ" crossorigin="anonymous">
  <link rel="stylesheet" href="https://unpkg.com/purecss@2.0.3/build/grids-responsive-min.css">
  <link rel="stylesheet" href="/assets/styles.css">
  
  <script type="text/x-mathjax-config">
MathJax.Hub.Config({
    TeX: {
        equationNumbers: {
            autoNumber: "AMS"
        }
    },
    tex2jax: {
        inlineMath: [ ['$','$'] ],
        displayMath: [ ['$$','$$'] ],
        processEscapes: true,
    }
});
</script>
<script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML">
</script>

  <!-- Begin Jekyll SEO tag v2.8.0 -->
<title>使用SV完成主成分分析 | kkmonster</title>
<meta name="generator" content="Jekyll v3.10.0" />
<meta property="og:title" content="使用SV完成主成分分析" />
<meta name="author" content="Feng kunjiang" />
<meta property="og:locale" content="en" />
<meta name="description" content="主成分分析(Principal component analysis)是一种降维的方法， 将高维特征映射到低维空间，也就是如果一个记录有很多特征，那么我们需要找到一组更少的特征来表示这众多的特征，这些更少的特征就称为主成分。" />
<meta property="og:description" content="主成分分析(Principal component analysis)是一种降维的方法， 将高维特征映射到低维空间，也就是如果一个记录有很多特征，那么我们需要找到一组更少的特征来表示这众多的特征，这些更少的特征就称为主成分。" />
<link rel="canonical" href="https://kunmonster.github.io//2025/04/11/%E4%BD%BF%E7%94%A8SV%E5%AE%8C%E6%88%90%E4%B8%BB%E6%88%90%E5%88%86%E5%88%86%E6%9E%90/" />
<meta property="og:url" content="https://kunmonster.github.io//2025/04/11/%E4%BD%BF%E7%94%A8SV%E5%AE%8C%E6%88%90%E4%B8%BB%E6%88%90%E5%88%86%E5%88%86%E6%9E%90/" />
<meta property="og:site_name" content="kkmonster" />
<meta property="og:type" content="article" />
<meta property="article:published_time" content="2025-04-11T20:30:00+08:00" />
<meta name="twitter:card" content="summary" />
<meta property="twitter:title" content="使用SV完成主成分分析" />
<meta name="google-site-verification" content="xxxxx" />
<script type="application/ld+json">
{"@context":"https://schema.org","@type":"BlogPosting","author":{"@type":"Person","name":"Feng kunjiang","url":"https://github.com/kunmonster/"},"dateModified":"2025-04-11T20:30:00+08:00","datePublished":"2025-04-11T20:30:00+08:00","description":"主成分分析(Principal component analysis)是一种降维的方法， 将高维特征映射到低维空间，也就是如果一个记录有很多特征，那么我们需要找到一组更少的特征来表示这众多的特征，这些更少的特征就称为主成分。","headline":"使用SV完成主成分分析","mainEntityOfPage":{"@type":"WebPage","@id":"https://kunmonster.github.io//2025/04/11/%E4%BD%BF%E7%94%A8SV%E5%AE%8C%E6%88%90%E4%B8%BB%E6%88%90%E5%88%86%E5%88%86%E6%9E%90/"},"url":"https://kunmonster.github.io//2025/04/11/%E4%BD%BF%E7%94%A8SV%E5%AE%8C%E6%88%90%E4%B8%BB%E6%88%90%E5%88%86%E5%88%86%E6%9E%90/"}</script>
<!-- End Jekyll SEO tag -->
<link type="application/atom+xml" rel="alternate" href="https://kunmonster.github.io//feed.xml" title="kkmonster" />
</head>


  <body>

    <div class="container pure-g"><div class="sidebar-left pure-u-1 pure-u-md-1-4" style="background-color: rgb(40, 73, 77); color: rgb(255, 255, 255); background-image: url(/bg.jpg);"><header class="masthead">
  <div class="avatar"><img src="/avatar.jpg" class="avatar-image" alt="Feng kunjiang">
    </div>
  <div class="masthead-title">
    <a href="/" title="Home">kkmonster</a>
  </div>
  <div class="masthead-tagline">
    <small>生命中有万事的可能</small>
  </div><nav class="navigation">
    <ul class="navigation-list"><li class="navigation-item">
        <a onclick="sessionStorage.setItem('forceCheckScroll', 'true')" href="/">Blog</a>
      </li><li class="navigation-item">
        <a onclick="sessionStorage.setItem('forceCheckScroll', 'true')" href="/about/">About</a>
      </li></ul>
  </nav><div class="social pure-menu pure-menu-horizontal">
      <ul class="social-icons pure-menu-list">
      <li class="pure-menu-item">
          <a class="social-icon pure-menu-link" href="mailto://fengkunjiang@stu.ynu.edu.cn">
            <i class="fas fa-envelope" title="Email"></i>
          </a>
        </li><li class="pure-menu-item">
          <a class="social-icon pure-menu-link" href="https://github.com/kunmonster">
            <i class="fab fa-github" title="Github"></i>
          </a>
        </li>
      </ul>
    </div>
</header>
</div>

      <div class="content pure-u-1 pure-u-md-1-2"><main>
  <article class="post">
  <h1 class="post-title">使用SV完成主成分分析</h1>
  <div class="post-meta"><time datetime="2025-04-11T20:30:00+08:00" itemprop="datePublished">11 Apr 2025</time></div>

  <p>主成分分析(Principal component analysis)是一种降维的方法， 将高维特征映射到低维空间，也就是如果一个记录有很多特征，那么我们需要找到一组更少的特征来表示这众多的特征，这些更少的特征就称为主成分。</p>

<p>在群体基因组中，我们经常使用PCA来查看群体间的关系，以及群体内分层的现象等，大部分时间都是使用SNP来进行分析，也就是各群体间共有的SNP位点，对他们的基因型进行主成分分析。</p>

<p>本文将使用结构变异(Structural variation)进行主成分分析。</p>

<h2 id="数据准备">数据准备</h2>
<hr />

<p>在进行主成分分析之前，我们需要准备一个VCF文件，其中包含每个变异的相关信息，以及每个样本的基因型信息。</p>

<p>这里有个小坑，我的结构变异是分群体进行calling的，需要先得到群体间的公共位点才行，但是结构变异的断点本身就是不准确的，所以这一步还是很困难的。</p>

<p>我的处理步骤如下：</p>

<ol>
  <li>提取每个群体的位点</li>
</ol>

<p>对每个群体提取所有变异位点，包含染色体号和变异起始位置</p>

<div class="language-shell highlighter-rouge"><div class="highlight"><pre class="highlight"><code>bcftools query <span class="nt">-f</span> <span class="s1">'%CHROM\t%POS\n'</span> your_vcf_file_1 | <span class="nb">sort</span> <span class="nt">-V</span> <span class="nt">-k1</span>,1 <span class="nt">-k2</span>,2n <span class="o">&gt;</span> site1.list
bcftools query <span class="nt">-f</span> <span class="s1">'%CHROM\t%POS\n'</span> your_vcf_file_2 | <span class="nb">sort</span> <span class="nt">-V</span> <span class="nt">-k1</span>,1 <span class="nt">-k2</span>,2n <span class="o">&gt;</span> site2.list
bcftools query <span class="nt">-f</span> <span class="s1">'%CHROM\t%POS\n'</span> your_vcf_file_3 | <span class="nb">sort</span> <span class="nt">-V</span> <span class="nt">-k1</span>,1 <span class="nt">-k2</span>,2n <span class="o">&gt;</span> site3.list
</code></pre></div></div>

<p><em>注意：上面命令中sort -V是按自然顺序排序，如果你文件中的染色体名是有英文字母的，如NCBI的NCxxxxx那么可以使用，如果你的染色体名字直接是数字那么就不需要加该选项</em></p>

<ol>
  <li>计算群体的共有位点</li>
</ol>

<div class="language-shell highlighter-rouge"><div class="highlight"><pre class="highlight"><code>
<span class="nb">awk</span> <span class="s1">'NR==FNR {pos[$0]; next} ($0 in pos)'</span> site1.list site2.list site3.list <span class="o">&gt;</span> common.list

</code></pre></div></div>

<ol>
  <li>提取每个群体共有位点的变异</li>
</ol>

<div class="language-shell highlighter-rouge"><div class="highlight"><pre class="highlight"><code>
bcftools view <span class="nt">-T</span> common.list <span class="nt">-O</span> z <span class="nt">-o</span> population_1.vcf.gz vcf_1
bcftools view <span class="nt">-T</span> common.list <span class="nt">-O</span> z <span class="nt">-o</span> population_2.vcf.gz vcf_2
bcftools view <span class="nt">-T</span> common.list <span class="nt">-O</span> z <span class="nt">-o</span> population_3.vcf.gz vcf_3

</code></pre></div></div>

<ol>
  <li>使用bcftools进行合并（同时使用bcftools计算出AF,MAF,MISSING RATE）</li>
</ol>

<div class="language-shell highlighter-rouge"><div class="highlight"><pre class="highlight"><code>
bcftools merge population_1.vcf.gz population_2.vcf.gz population_3.vcf.gz <span class="nt">-O</span> z <span class="nt">-o</span> common_var.vcf.gz

bcftools +fill-tags common_var.vcf.gz <span class="nt">--</span> <span class="nt">-t</span> AF | bcftools +fill-tags <span class="nt">--</span> <span class="nt">-t</span> MAF | bcftools +fill-tags <span class="nt">-O</span> z <span class="nt">-o</span> common_var_with_info.vcf.gz <span class="nt">--</span> <span class="nt">-t</span> F_MISSING

</code></pre></div></div>

<ol>
  <li>将vcf文件中的ALT字段全部替换成PLINK可接受的输入</li>
</ol>

<div class="language-shell highlighter-rouge"><div class="highlight"><pre class="highlight"><code>
<span class="c"># 此处假设vcf文件中ALT字段为&lt;xxxxx&gt;类型</span>
<span class="c"># 使用sed替换</span>

bcftools view common_var_with_info.vcf.gz | <span class="nb">sed</span> <span class="nt">-E</span> <span class="s1">'/^#/! s/&lt;[^&gt;]+&gt;/SV/g'</span> <span class="o">&gt;</span> common_var_with_info.vcf

bgzip common_var_with_info.vcf
tabix common_var_with_info.vcf.gz

</code></pre></div></div>

<ol>
  <li>将vcf文件转换成PLINK的标准输入（使用PLINK2）</li>
</ol>

<div class="language-shell highlighter-rouge"><div class="highlight"><pre class="highlight"><code>plink2 <span class="nt">--vcf</span> common_var_with_info.vcf.gz <span class="se">\</span>
<span class="nt">--geno</span> 0.05 <span class="se">\ </span>  <span class="c"># 每个变异的missing rate &lt; 5%</span>
<span class="nt">--maf</span> 0.05 <span class="se">\ </span>   <span class="c"># 每个变异maf &gt; 0.05</span>
<span class="nt">--hwe</span> 1e-6 <span class="se">\ </span>   <span class="c"># 满足哈代温伯格平衡</span>
<span class="nt">--freq</span> <span class="se">\ </span>       <span class="c"># 加载基因型频率</span>
<span class="nt">--allow-extra-chr</span> <span class="se">\ </span><span class="c"># 由于我使用NCBI refseq的染色体ID,所以加上该选项,如果是数字染色体名则不需要</span>
<span class="nt">--make-bed</span> <span class="se">\</span>
<span class="nt">--out</span> plink/plink_input <span class="c"># 将结果存放到plink目录下,结果文件以plink_input为前缀</span>
</code></pre></div></div>

<h2 id="主成分分析及可视化">主成分分析及可视化</h2>
<hr />

<p>在将数据转换成PLINK标准输入后，使用PLINK继续进行主成分分析，并且使用R以及ggplot2包进行可视化。</p>

<ol>
  <li>计算主成分</li>
</ol>

<div class="language-shell highlighter-rouge"><div class="highlight"><pre class="highlight"><code>
plink2 <span class="nt">--bfile</span> plink_input <span class="nt">--pca</span> 10 <span class="nt">-out</span> pca_res

</code></pre></div></div>

<p>计算主成分后将会获得两个文件</p>

<ul>
  <li>pca_res.eigenval （每个主成分的比例，无表头）</li>
  <li>pca_res.eigenvec（样本为行，主成分为列，有表头）</li>
</ul>

<ol>
  <li>进行可视化</li>
</ol>

<p>将上述两个文件，使用R打开，注意，我这里有三个群体，还需要手动在pca_res.eigenvec中加一下群体信息</p>

<div class="language-R highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">eigenvec</span><span class="o">&lt;-</span><span class="n">read.csv</span><span class="p">(</span><span class="s1">'pca_res.eigenvec'</span><span class="p">,</span><span class="w"> </span><span class="n">header</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">,</span><span class="n">sep</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"\t"</span><span class="p">)</span><span class="w">

</span><span class="n">eigenval</span><span class="o">&lt;-</span><span class="n">read.table</span><span class="p">(</span><span class="s2">"pca_res.eigenval"</span><span class="p">,</span><span class="w"> </span><span class="n">header</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">FALSE</span><span class="p">)</span><span class="w">



</span><span class="c1"># 去掉第一列并加上分组信息</span><span class="w">

</span><span class="n">library</span><span class="p">(</span><span class="n">dplyr</span><span class="p">)</span><span class="w">
</span><span class="n">eigenvec</span><span class="o">&lt;-</span><span class="n">mutate</span><span class="p">(</span><span class="n">eigenvec</span><span class="p">[</span><span class="m">-1</span><span class="p">],</span><span class="n">group</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">substr</span><span class="p">(</span><span class="n">IID</span><span class="p">,</span><span class="m">1</span><span class="p">,</span><span class="m">2</span><span class="p">))</span><span class="w">

</span><span class="n">pca_data</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">data.frame</span><span class="p">(</span><span class="w">
  </span><span class="n">Sample</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">eigenvec</span><span class="p">[,</span><span class="w"> </span><span class="m">1</span><span class="p">],</span><span class="w"> </span><span class="c1"># 第一列样本信息</span><span class="w">
  </span><span class="n">Group</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">factor</span><span class="p">(</span><span class="n">eigenvec</span><span class="o">$</span><span class="n">group</span><span class="p">),</span><span class="w">
  </span><span class="n">PC1</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">eigenvec</span><span class="p">[,</span><span class="w"> </span><span class="m">2</span><span class="p">],</span><span class="w"> </span><span class="c1"># 主成分1</span><span class="w">
  </span><span class="n">PC2</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">eigenvec</span><span class="p">[,</span><span class="w"> </span><span class="m">3</span><span class="p">]</span><span class="w">  </span><span class="c1"># 主成分2</span><span class="w">
</span><span class="p">)</span><span class="w">

</span><span class="c1"># 计算方差解释比例</span><span class="w">
</span><span class="n">var_explained</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">eigenval</span><span class="o">$</span><span class="n">V1</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="nf">sum</span><span class="p">(</span><span class="n">eigenval</span><span class="o">$</span><span class="n">V1</span><span class="p">)</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="m">100</span><span class="w">

</span><span class="n">library</span><span class="p">(</span><span class="s1">'ggplot2'</span><span class="p">)</span><span class="w">
</span><span class="c1"># 绘制PCA图</span><span class="w">
</span><span class="n">ggplot</span><span class="p">(</span><span class="n">pca_data</span><span class="p">,</span><span class="w"> </span><span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">PC1</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">PC2</span><span class="p">,</span><span class="w"> </span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">Group</span><span class="p">))</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">geom_point</span><span class="p">(</span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">3</span><span class="p">,</span><span class="w"> </span><span class="n">alpha</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.7</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">labs</span><span class="p">(</span><span class="w">
    </span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">paste0</span><span class="p">(</span><span class="s2">"PC1 ("</span><span class="p">,</span><span class="w"> </span><span class="nf">round</span><span class="p">(</span><span class="n">var_explained</span><span class="p">[</span><span class="m">1</span><span class="p">],</span><span class="w"> </span><span class="m">2</span><span class="p">),</span><span class="w"> </span><span class="s2">"%)"</span><span class="p">),</span><span class="w">
    </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">paste0</span><span class="p">(</span><span class="s2">"PC2 ("</span><span class="p">,</span><span class="w"> </span><span class="nf">round</span><span class="p">(</span><span class="n">var_explained</span><span class="p">[</span><span class="m">2</span><span class="p">],</span><span class="w"> </span><span class="m">2</span><span class="p">),</span><span class="w"> </span><span class="s2">"%)"</span><span class="p">),</span><span class="w">

  </span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">stat_ellipse</span><span class="p">(</span><span class="n">level</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.95</span><span class="p">,</span><span class="w"> </span><span class="n">linetype</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">2</span><span class="p">,</span><span class="w"> </span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.5</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">  </span><span class="c1"># 添加95%置信椭圆</span><span class="w">
  </span><span class="n">theme_bw</span><span class="p">()</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">theme</span><span class="p">(</span><span class="w">
    </span><span class="n">legend.position</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"top"</span><span class="p">,</span><span class="w">
    </span><span class="n">plot.title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_text</span><span class="p">(</span><span class="n">hjust</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.5</span><span class="p">)</span><span class="w">
  </span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<p>然后就会得到如下PCA图</p>

<p><img src="https://raw.githubusercontent.com/kunmonster/note_pic/main/note/202504012006653.png" alt="image.png" /></p>

<h2 id="最后">最后</h2>
<hr />

<p>使用结构变异来做PCA分析还是有很多小坑，特别是PLINK对SV支持不是很好，ALT列不能识别，需要手动替换</p>


  
    
      <div class="post-tags-section">
  <i class="post-tags-icon fas fa-tags"></i>
  <ul class="post-tags"><li>
        <a class="post-tag" href="/tags/#sv">sv</a></li><li>
        <a class="post-tag" href="/tags/#pca">pca</a></li></ul>
</div>

  

  
</article>


<aside class="related">
  <h2>Related posts</h2>
  <ul class="related-posts">
    
      <li>
        <a href="/2025/04/14/Circos_draw_histogram/">
          Drawing variation histogram by Circos
          <small><time datetime="2025-04-14T22:00:00+08:00">14 Apr 2025</time></small>
        </a>
      </li>
    
      <li>
        <a href="/2025/03/16/DeepsvFilter-%E5%9F%BA%E4%BA%8E%E6%B7%B1%E5%BA%A6%E5%AD%A6%E4%B9%A0%E5%AE%8C%E6%88%90%E7%BB%93%E6%9E%9C%E5%8F%98%E5%BC%82%E8%BF%87%E6%BB%A4/">
          DeepsvFilter:基于深度学习完成结果变异过滤
          <small><time datetime="2025-03-16T17:11:00+08:00">16 Mar 2025</time></small>
        </a>
      </li>
    
      <li>
        <a href="/2024/11/16/IGV%E5%8F%AF%E8%A7%86%E5%8C%96BAM%E6%96%87%E4%BB%B6%E5%B8%AE%E5%8A%A9%E7%90%86%E8%A7%A3%E7%BB%93%E6%9E%84%E5%8F%98%E5%BC%82/">
          IGV可视化BAM文件帮助理解结构变异
          <small><time datetime="2024-11-16T17:42:00+08:00">16 Nov 2024</time></small>
        </a>
      </li>
    
  </ul>
</aside>


</main>

<footer class="footer"><small>
    &copy; 2023&nbsp;-&nbsp;2025 <a href="https://github.com/kunmonster/">Feng kunjiang</a>. All rights reserved.
  </small>
</footer>
</div>
      <div class="sidebar-right pure-u-1 pure-u-md-1-4">
<div  class="toc-wrapper">
  <h2 class="toc-title">Table of Contents</h2>
    <nav class="toc-nav">
      <ul class="toc">
  <li><a href="#数据准备">数据准备</a></li>
  <li><a href="#主成分分析及可视化">主成分分析及可视化</a></li>
  <li><a href="#最后">最后</a></li>
</ul>

  </nav>
</div>

</div>
    </div>

    <script async src="https://use.fontawesome.com/releases/v5.0.12/js/all.js"></script><script>
  function strip(str, remove) {
    while (str.length > 0 && remove.indexOf(str.charAt(0)) != -1) {
      str = str.substr(1);
    }
    while (str.length > 0 && remove.indexOf(str.charAt(str.length - 1)) != -1) {
      str = str.substr(0, str.length - 1);
    }
    return str;
  }

  function scroll() {
    console.log('scroll');
    window.scrollTo({
      left: 0, 
      top: window.innerHeight,
      behavior: 'smooth'
    });
    sessionStorage.removeItem('forceCheckScroll');
  }

  const forceCheckScroll = sessionStorage.getItem('forceCheckScroll') === 'true';
  const checkScroll = strip(window.location.pathname, '/') !== strip('', '/');

  if (forceCheckScroll || checkScroll) {
    const maxWidth = "(max-width: 48rem)";
    const result = window.matchMedia(maxWidth);
    if (result.matches) {
      scroll();
    } else {
      result.addListener((match) => {
        if (match.media == maxWidth) {
          if (match.matches) {
            scroll();
          }
        }
      });
    }
  }
</script>
</body>
</html>
